bundle/edu.gemini.spModel.core/src/main/scala/edu/gemini/spModel/core/ProperMotion.scala

```diff
@@ -11,16 +15,126 @@ import Scalaz._
 case class ProperMotion(
   deltaRA: RightAscensionAngularVelocity,
   deltaDec: DeclinationAngularVelocity,
-  epoch: Epoch = Epoch.J2000)
+  epoch: Epoch = Epoch.J2000) {
+
+  /** Coordinates at the current time. */
+  def calculateAt(t: Target, i: Instant): Coordinates =
+    plusYears(t, epoch.untilInstant(i))
+
+  /** Coordinates `elapsedYears` fractional epoch-years after `epoch`. */
+  def plusYears(t: Target, elapsedYears: Double): Coordinates = {
+    val baseCoordinates = t.coords(None).getOrElse(Coordinates.zero)
+
+    // TODO: Is this calculated correctly?
+    val properVelocity: Offset = Offset(OffsetP(Angle.fromDegrees(deltaRA.velocity.toDegreesPerYear)),
+                                        OffsetQ(Angle.fromDegrees(deltaDec.velocity.toDegreesPerYear)))
+
+    val radialVelocity: Double = {
+      val redshift: Option[Redshift] = Target.redshift.get(t).flatten
+      redshift.getOrElse(Redshift.zero).toRadialVelocity.toKilometersPerSecond
+    }
+    val parallax = Parallax.mas.get(Target.parallax.get(t).flatten.getOrElse(Parallax.zero)) / 1000000.0
+    ProperMotion.properMotionCalculator(baseCoordinates,
+                                        Epoch.JulianLengthOfYear,
+                                        properVelocity,
+                                        radialVelocity,
+                                        parallax,
+                                        elapsedYears)
+  }
+}
 
 object ProperMotion {
+  // Some constants we need
+  val AstronomicalUnit:    Long   = 149597870660L
+  private val secsPerDay:  Double = 86400.0
+  private val auPerKm:     Double = 1000.0 / AstronomicalUnit.toDouble
+  private val radsPerAsec: Double = Angle.fromArcsecs(1).toArcsecs
+  private val TwoPi:       Double = 6.283185307179586476925286766559
+
+  private type Vec2 = (Double, Double)
+  private type Vec3 = (Double, Double, Double)
+
+  // Scalar multiplication and addition for Vec3.
+  // TODO: Does scalaz must provide monoids for addition and multiplication?
```

Comment:
I think there are but you may need to import them explicitly

Reply:
I'm hoping @tpolecat will know.

Reply:
The issue seems to be with [floating point representation](https://github.com/scalaz/scalaz/issues/334), not with the tuples.  I wouldn't worry about it, this seems fine.